articles/102_dependency_groups.md

```diff
@@ -0,0 +1,169 @@
+It will need all package manifests available from `rosdistro` to resolve a dependency group to a set of packages.
+
+For the `satisfy-all` semantic it can simply generate dependencies on all packages from the group.
+For the `satisfy-at-least-one` semantic the release tool could either use a platform specific mechanism like `provides` in Debian control files or simply expand the group to all packages (same as `satisfy-all`).
+This design document doesn't aim to specify what the "best" behavior for the release tool is since the decision is likely platform dependent.
+
+Since the release tool expands the dependency groups when being invoked, packages which join the group after that release won't be used until the package defining the group dependency is being re-released.
+
+### Buildfarm
+
+Since the buildfarm uses the topological order to trigger downstream binary jobs it needs to consider the dependency groups.
+The same interpretation of the `satify` semantic as for `bloom` applies.
+
+### rosinstall_generator
+
+The `rosinstall_generator` already utilized all manifests from the `rosdistro` to decide which packages / repositories should be fetched.
+This design document doesn't aim to specify what the "best" behavior for this tool is.
+It could either match the behavior of the build tool or of the release tool or even something different.
+
+## Information in the Manifest
+
+To provide the necessary metadata for the build and release tools the following information must be provided by the manifest:
+
+* (A) A specific package declares a group dependency and uses a name to identify the group.
+  The dependency can be of any type specified in [REP 140](http://www.ros.org/reps/rep-0140.html): e.g. `build`, `exec`, `test`, `doc`, etc.
+
+* (B) The semantic of the group dependency needs to be defined (in case (1) `satisfy-at-least-one`, in case (2) `satify-all`).
+
+* (C) A list of preferred / excluded packages can be enumerated for each dependency group.
+
+* (D) Any package can declare that it is part of a group by name.
+
+### Possible Format Extensions
+
+The group dependency (A) could be specified in two ways:
+
+* (A.1) Using the existing `*_depend` tags and adding an attribute to the tag to make it a group dependency.
+* (A.2) Defining new `*_group_depend` tags which specifically identify group dependencies.
+
+Considering the existing API in `catkin_pkg` which exposes the different dependencies as members of the `Package` class (A.1) would require existing code to distinguish the type of the dependency (group vs. no-group).
+On the other hand (A.2) will simply add additional members which doesn't require existing code to change until it wants to support handling group dependencies.
+
+The semantic of the group dependency (B) can be modeled as an attribute of the group dependency tag.
+
+Each preferred / excluded package (C) name can be listed in child-tags of the dependency group to keep the information atomic.
```

Comment:
Exclusions and preferences are mentioned here and above, but there are no examples in the Example section, and I didn't see a mention of it in REP 0149.  It would be nice to see examples of that.

Reply:
For the scope of this document I would rather keep it as-is with a text description. The XML could look like this and will be specified in the REP once we have a consensus about the direction:

```
<group_depend>
  group_name
  <prefer>try_this_first</prefer>
  <prefer>try_this_second</prefer>
  <exclude>do_not_use_that</exclude>
</group_depend>
```